source/expressions.tex

```diff
@@ -661,6 +661,10 @@
 type, the conversion copy-initializes the result object from
 the glvalue.
 
+\item Otherwise, if $E$ is a constant expression that designates an object
+that is usable in constant expressions ([expr.const]),
+the result is the value of that object.
```

Comment:
This seems much broader than the rule in /2.2. This would also remove the access in:

```
struct X { int a; };
constexpr X x = {1};
int n = (&x)->a;
```

... which currently does result in an odr-use of `x` and an access. Changing that doesn't seem editorial.

Can we make the words here more directly refer back to the /2.2 case somehow? Maybe something like:

```
\item Otherwise, if $E$ is potentially evaluated and the object designated by $E$ is not accessed, as specified above, the designated object is necessarily usable in constant expressions ([expr.const]) and the result is the value of that object.
```

(I don't like "is necessarily" here. Do we have another way to say "must be" without making ISO angry?)

Reply:
Or maybe we could inline /2 into these bullets?

Reply:
This also has interactions with the very definition of "lvalue":

```
struct S
{
     static const int a = 1;
};

int main()
{
     return S::a;  // no odr-use; no definition of S::a needed
}
```

[intro.object] p1 "An object is created by a definition (6.2) ..."

[basic.lval] p1 "A glvalue is an expression whose evaluation determines the identity of an object or function.

How can you identify an object that wasn't ever created (because there is no definition)?

http://nolp.dhl.de/nextt-online-public/set_identcodes.do?lang=de&idc=579347005106

Reply:
@jensmaurer I think you pasted a wrong link

Reply:
I think it works best to integrate p2 into the "result" bullets.

Reply:
Maybe say `the result is the value of \tcode{x}'s initializer.` in non-odr-used variable case? Variable is not an object after all, it doesn't have value. And this

Reply:
BTW
```c++
struct S
{
	int m;
};

int main()
{
	&S::m;
}
```
[expr.prim.id.qual] says that a _qualified-id_ denoting data member is an lvalue, but I don't think `S::m` in `&S::m` can denote an object. Even though we have pointer-to-member, IIUC `S::m` is still an lvalue there, which makes me feel uncomfortable. Should this be a Core issue? 🤔

Reply:
*Maybe say the result is the value of \tcode{x}'s initializer*. A variable that is usable in constant expressions does not always has an initializer 
> A variable or temporary object o is constant-initialized if. 
>> - either it has an initializer or **its default-initialization results in some initialization being performed** 
````cpp
struct A{
    int v = 1;
};
````